### ISSUE TITLE
[BUG] Dependencies of published integration tests jar are missing

### ISSUE THREAD

**Issue by jlowe** [NVIDIA/spark-rapids#3934]
Attempting to download the integration tests jar as a dependency fails with missing dependencies:
```
$ mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-integration-tests_2.12 -Dversion=21.08.0 -Dpackaging=jar
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:2.8:get (default-cli) on project rapids-4-spark-parent: Couldn't download artifact: Missing:
[ERROR] ----------
[ERROR] 1) com.nvidia:rapids-4-spark-tests_2.12:test-jar:tests:21.08.0
[ERROR] 
[ERROR]   Try downloading the file manually from the project website.
[ERROR] 
[ERROR]   Then, install it using the command: 
[ERROR]       mvn install:install-file -DgroupId=com.nvidia -DartifactId=rapids-4-spark-tests_2.12 -Dversion=21.08.0 -Dclassifier=tests -Dpackaging=test-jar -Dfile=/path/to/file
[ERROR] 
[ERROR]   Alternatively, if you host your own repository you can deploy the file there: 
[ERROR]       mvn deploy:deploy-file -DgroupId=com.nvidia -DartifactId=rapids-4-spark-tests_2.12 -Dversion=21.08.0 -Dclassifier=tests -Dpackaging=test-jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR] 
[ERROR]   Path to dependency: 
[ERROR]   	1) org.apache.maven.plugins:maven-downloader-plugin:jar:1.0
[ERROR]   	2) com.nvidia:rapids-4-spark-integration-tests_2.12:jar:21.08.0
[ERROR]   	3) com.nvidia:rapids-4-spark-tests_2.12:test-jar:tests:21.08.0
[ERROR] 
[ERROR] 2) com.nvidia:rapids-4-spark-udf-examples_2.12:jar:21.08.0
[ERROR] 
[ERROR]   Try downloading the file manually from the project website.
[ERROR] 
[ERROR]   Then, install it using the command: 
[ERROR]       mvn install:install-file -DgroupId=com.nvidia -DartifactId=rapids-4-spark-udf-examples_2.12 -Dversion=21.08.0 -Dpackaging=jar -Dfile=/path/to/file
[ERROR] 
[ERROR]   Alternatively, if you host your own repository you can deploy the file there: 
[ERROR]       mvn deploy:deploy-file -DgroupId=com.nvidia -DartifactId=rapids-4-spark-udf-examples_2.12 -Dversion=21.08.0 -Dpackaging=jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR] 
[ERROR]   Path to dependency: 
[ERROR]   	1) org.apache.maven.plugins:maven-downloader-plugin:jar:1.0
[ERROR]   	2) com.nvidia:rapids-4-spark-integration-tests_2.12:jar:21.08.0
[ERROR]   	3) com.nvidia:rapids-4-spark-udf-examples_2.12:jar:21.08.0
[ERROR] 
[ERROR] ----------
[ERROR] 2 required artifacts are missing.
```

We need to do at least one of the following:
- Publish the missing artifacts
- Fix the dependencies to remove the missing artifacts (if they indeed should be removed)
- Stop publishing this broken artifact

**Comment by res-life** [comment]
Issue #3932 will remove aggregator dependency from this integration module, and will remove tests module, so it's better to fix  #3932 first.

**Comment by res-life** [comment]
@jlowe
The current version is working fine if we specify all the necessary parameters.
```
// Will download successfully if providing the classifier and packageing parameters:
mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-integration-tests_2.12 -Dversion=22.02.0-SNAPSHOT -Dpackaging=tar.gz -Dclassifier=pytest
mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-integration-tests_2.12 -Dversion=22.02.0-SNAPSHOT -Dclassifier=spark301
mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-sql_2.12 -Dversion=22.02.0-SNAPSHOT -Dpackaging=jar -Dclassifier=spark301

// this will fail without classifier parameter
mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-sql_2.12 -Dversion=22.02.0-SNAPSHOT -Dpackaging=jar
```

Are there any other considerations?

**Comment by jlowe** [comment]
> The current version is working fine if we specify all the necessary parameters.

Those commands are only working for you because it is picking up artifacts that you built locally and installed in your local Maven repository (or you are using Maven settings that add an internal repository that has these artifacts available).  It will fail on a fresh Maven repository (e.g.: add `-Dmaven.repo.local=/tmp/mynewlocalrepo` to any of the command-lines above) with no additional Maven settings that add non-public repositories.

**Comment by res-life** [comment]
Currently, we do not public the internal jars like sql-plugin, shuffle, see the https://repo.maven.apache.org/maven2/com/nvidia, only published dist, tool, integration.
Integration_tests depends on the aggregator, udf-example and the tests modules. I checked, all the dependencies are necessary. 
The aggregator depends on sql-plugin and shuffle.
"mvn dependency:get" will pull all the dependencies and some of the dependencies are not published.
Is it necessary to publish all the dependencies? 
By sepcifing -Dtransitive will only download the solo jar without the dependencies.
see the following:
	mvn dependency:get -B -Ddest=snap -DgroupId=com.nvidia -DartifactId=rapids-4-spark-integration-tests_2.12 -Dversion=21.10.0 -Dclassifier=spark301 -Dmaven.repo.local=/tmp/.m2 -Dtransitive=false

**Comment by jlowe** [comment]
> Is it necessary to publish udf-example and tests test-jar?

The plan is to move the udf-examples into a separate repository like [spark-rapids-examples](https://github.com/NVIDIA/spark-rapids-examples).  Currently that project isn't a very good example project, as it depends on the rapids-4-spark-sql jar rather than the dist jar.  Users trying to copy the example will get a rude surprise, as the rapids-4-spark-sql jar is not published, nor do we intend to publish it as it contains classes user code should not access directly (and those classes are hidden in the dist jar).

The problem with publishing the tests jar is that it just moves the problem being described here from the integration tests jar to the tests jar.  The tests jar will then have dependencies that are unpublished in Maven Central, and I don't think we want to publish all those dependencies (per discussion above about classes we do not want to be directly accessed).

> Use -Dtransitive

I'm -1 on this being considered a real "fix."  IMO that is a workaround to a bug in the published dependencies of the artifact, so it does not address the problem being described here.  It just provides a workaround for those dealing with this bug.

AFAICT the only reason we're publishing this artifact is as a convenience to get this to our nightly test pipelines.  So one option is to stop publishing this to the public repo and only publish to an internal repo that is accessed by the nightly pipelines.

> Move the following to a project

Moving these out of the tests module, which has problematic dependencies for public publishing per the discussion in #3932, seems like it could be a good idea.  That should be able to be cleanly published as a separate module or moved to the integration tests module.

To sum up, we either need to stop publishing this publicly and just use the internal repo for out pipelines or work on the following (which I think we should do even if we stop publishing the integration tests jar to Maven Central):
- Getting the udf-examples moved to a separate repository along with its corresponding tests.  We will need to setup nightly test pipelines for it and we should start publishing that artifact, as it has useful UDFs that users could use as-is.
- Move the code identified above out of the tests module to make it easier to use without pulling in the problematic dependencies of the tests module.

**Comment by res-life** [comment]
After discussion:
    Stop publishing the integration tests jar to Maven Central.  @NvTimLiu  Please remove it.
    Remove the following dependency in the integration_tests module pom.xml. @NvTimLiu
```
//   it's not needed.
<!-- use aggregator jar because accessing internal classes -->  
        <dependency>
            <groupId>com.nvidia</groupId>
            <artifactId>rapids-4-spark-aggregator_${scala.binary.version}</artifactId>
            <version>${project.version}</version>
            <classifier>${spark.version.classifier}</classifier>
            <scope>test</scope>
        </dependency>
```

[**blocking**] Move the udf-examples module to the spark-rapids-examples repository; Publish the udf-examples snapshot jar and rapids-4-spark snapshot jar to Maven Central; Let udf-examples depends on rapids-4-spark; Keep the rapids_udf_test.py in the integration tests. This step is blocked because rapids-4-spark snapshot jar is not published to Maven Central currently.

@jlowe any other info?

**Comment by res-life** [comment]
The Scala tests in the integration-tests module depend on SparkQueryCompareTestSuite resides in the test module.
Attemped to extract SparkQueryCompareTestSuite to a separate project, and then let the integration-tests module and the tests module both depend on the new project, but failed.
SparkQueryCompareTestSuite actually depends on RapidsConf and ShimLoader which are internal.
The dependency paths are:

    SparkQueryCompareTestSuite ->
        RapidsConf
        ExecutionPlanCaptureCallback ->
            PlanUtils ->
                ShimLoader.loadClass ->
                   ShimLoader.getShimClassLoader()

**Comment by jlowe** [comment]
All of this is a good start but not sufficient.  As long as the integration tests jar depends on the tests jar, it will also be transitively depending on the aggregation jar.

> The Scala tests in the integration-tests module depend on SparkQueryCompareTestSuite resides in the test module.

I think those tests should be moved to the tests module then.  There are many tests in the tests module that are similarly constructed, so I don't see why these tests would require special treatment and be placed in the integration tests module.

**Comment by NvTimLiu** [comment]
Reopen, there are PRs for the issue

**Comment by NvTimLiu** [comment]
Reopen, there are PRs for the issue

**Comment by NvTimLiu** [comment]
Ropen

**Comment by res-life** [comment]
Fixed: 
1. Stop publishing the integration tests jar to Central Maven.
2. Remove the dependency from the integration module to the tests module.
